lib/Conversion/PolyToStandard/PolyToStandard.cpp

```diff
@@ -60,11 +66,51 @@ struct ConvertAdd : public OpConversionPattern<AddOp> {
 
   using OpConversionPattern::OpConversionPattern;
 
+  // Convert add lowers a poly.add operation to arith operations. A poly.add
+  // operation is defined within the polynomial ring. Coefficients are added
+  // element-wise as elements of the ring, so they are performed modulo the
+  // coefficient modulus.
+  //
+  // Modular addition is performed by sign extending both arguments and
+  // performing (N+1)-bit addition. Then, if z = (x + y) is the (N+1)-bit sum,
+  // we return
+  //   truncate(z >= mod ? z - mod : z)
   LogicalResult matchAndRewrite(
       AddOp op, OpAdaptor adaptor,
       ConversionPatternRewriter &rewriter) const override {
-    // TODO(https://github.com/google/heir/issues/104): implement
-    return failure();
+    auto type = adaptor.getLhs().getType();
+    auto elementTy =
+        type.cast<RankedTensorType>().getElementType().cast<IntegerType>();
+    auto extType = IntegerType::get(getContext(), elementTy.getWidth() + 1,
+                                    elementTy.getSignedness());
+    auto extShapeType = type.cast<ShapedType>().clone(extType);
+
+    mlir::Value extLhsOp = rewriter.create<arith::ExtUIOp>(
+        op.getLoc(), extShapeType, adaptor.getLhs());
+    mlir::Value extRhsOp = rewriter.create<arith::ExtUIOp>(
+        op.getLoc(), extShapeType, adaptor.getRhs());
+
+    auto addOp = rewriter.create<arith::AddIOp>(op.getLoc(), extShapeType,
```

Comment:
Coefficients in poly are by definition unsigned. Or rather, the coefficient ring is always Z/nZ, which I'm assuming means we're picking the representative in [0, n-1] for all our computations. So shouldn't all these ops use the unsigned variant? Also, if the semantics were signed then you'd want signed extension, not zero-extension. I think it's just easier to do everything unsigned.

Ditto for the tests that create polys with i64s. Maybe this means we should have a verifier that the coefficients used to construct a poly are always unsigned?

Reply:
On point 1: yep, agreed to use addui and unsigned arith ops.

> Maybe this means we should have a verifier that the coefficients used to construct a poly are always unsigned?

I'm not sure, if we want our type conversion to handle `from_coeff` by just dropping the op and assigning to the original tensor, then I don't think we can use unsigned ints. Most of the arith dialect ops want signless integers, despite being for unsigned ops. So for example, `arith.addui_extended` requires operands to be `signless-integer-like` - which explicitly disallows `ui64`. 

Maybe this is why you originally used `i64`?

Reply:
Yeah I ran into the same issue. Signless will have to be the way to go, and I think i32 counts as signless?